### PR TITLE
linkage_checker: skip broken linkage in Julia

### PR DESCRIPTION
**Pull request by cho-m**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Looking into a way to avoid Julia linkage errors since it uses its own dylib lookup logic which brew considers broken linkage.

Seemed to be easier approach rather than trying to inject unneeded RPATHs into the binary that Julia doesn't want.

Example for Julia:
```json
{
  "julia": "share/julia/compiled/**/*"
}
```

Need to try out a few more things and check on Linux code path.

### CODE REVIEW

**Review by MikeMcQuaid — APPROVED** [review]
Good idea!

**Comment by Bo98** [comment]
Can you talk about what's broken about Julia currently? We already have it shipped in homebrew-core with passing linkage.

I feel quite strongly that `brew linkage` should never return false positives, so if it is that is concerning. I know the RPATH part of it is debated (which is why it has never graduated from a warning) but the broken linkage one is another story.

We did have a `DSL` allowlist in the past but we removed it because we ended up fixing the issues.

Is there anything _except_ Julia that would use this? If it's only Julia we might as well hardcode in `share/julia/compiled` as that's an unusual path to be putting binaries anyway.

**Comment by cho-m** [comment]
> Can you talk about what's broken about Julia currently? We already have it shipped in homebrew-core with passing linkage.

We shipped it with broken linkage when check was just a warning - https://github.com/Homebrew/homebrew-core/pull/173754/commits/cb36837eca49200b31ff45e96d3ea01788a0b15d

The Linux broken linkage covers the same failure as macOS missing RPATHs. It's just that the missing RPATHs are detected by finding `@rpath`. We could classify Linux shared libraries with non-resolvable NEEDED libraries and without RPATH/RUNPATH as missing, but it could also just be broken.

---

> Is there anything _except_ Julia that would use this? If it's only Julia we might as well hardcode in `share/julia/compiled` as that's an unusual path to be installing binaries to anyway.

Not sure about others. We could hardcode in Julia. Right now, the dylibs in that directory are not intended to be used with native dynamic loader which is why they trigger linkage error. I think they are run through Julia's loader instead.

**Comment by cho-m** [comment]
> Not sure about Linux - I would have expected that to fail.

Currently it does fail so probably fixed (or at least elevated from warning), e.g.
* https://github.com/Homebrew/homebrew-core/pull/182774/files
* https://github.com/Homebrew/homebrew-core/pull/183057/files

---

> This linkage check runs even when you are running `brew install`. `audit_exceptions` is not available under the API so should not be used to make hard-failing linkage checks pass. Tests that cannot achieve 100% accuracy must be placed under strict mode or relocated to `FormulaCellarChecks` used by `brew audit`. So we should not add an allowlist to `broken_dylibs` as that is checked under non-strict mode.

I guess easiest path for now is the hardcoding of Julia. Moving it to `brew audit` may be better in long run.

**Review by carlocab — APPROVED** [review]
May as well add `&& !broken_dylibs_allowed?(file)` here too:

https://github.com/Homebrew/brew/blob/f0be97f59219c29e4c70decbd59cb04e24fbdf93/Library/Homebrew/linkage_checker.rb#L131

**Review by carlocab — APPROVED** [review]

**Comment by cho-m** [comment]
On topic of Julia but unrelated to this PR, it may be nice to detect symlinks to other formulae and use that to avoid some "Dependencies with no linkage". Julia does this for most dependencies thus we get:
```
Dependencies with no linkage:
  ca-certificates
  gmp
  libnghttp2
  libssh2
  mpfr
  openblas
  openlibm
```
```console
❯ eza -1 $(brew --prefix julia)/lib/julia/ | rg ' -> \.\.'
libatomic.1.dylib -> ../../../../../opt/gcc/lib/gcc/current/libatomic.1.dylib
libblastrampoline.dylib -> ../../../../../opt/libblastrampoline/lib/libblastrampoline.5.dylib
libcurl.dylib -> ../../../../../opt/curl/lib/libcurl.4.dylib
libgcc_s.1.1.dylib -> ../../../../../opt/gcc/lib/gcc/current/libgcc_s.1.1.dylib
libgfortran.5.dylib -> ../../../../../opt/gcc/lib/gcc/current/libgfortran.5.dylib
libgit2.dylib -> ../../../../../opt/libgit2/lib/libgit2.1.8.dylib
libgmp.dylib -> ../../../../../opt/gmp/lib/libgmp.10.dylib
libgomp.1.dylib -> ../../../../../opt/gcc/lib/gcc/current/libgomp.1.dylib
libmbedcrypto.dylib -> ../../../../../opt/mbedtls@2/lib/libmbedcrypto.7.dylib
libmbedtls.dylib -> ../../../../../opt/mbedtls@2/lib/libmbedtls.14.dylib
libmbedx509.dylib -> ../../../../../opt/mbedtls@2/lib/libmbedx509.1.dylib
libmpfr.dylib -> ../../../../../opt/mpfr/lib/libmpfr.6.dylib
libnghttp2.dylib -> ../../../../../opt/libnghttp2/lib/libnghttp2.14.dylib
libopenblas.dylib -> ../../../../../opt/openblas/lib/libopenblas.0.dylib
libopenlibm.dylib -> ../../../../../opt/openlibm/lib/libopenlibm.4.dylib
libpcre2-8.dylib -> ../../../../../opt/pcre2/lib/libpcre2-8.0.dylib
libquadmath.0.dylib -> ../../../../../opt/gcc/lib/gcc/current/libquadmath.0.dylib
libssh2.dylib -> ../../../../../opt/libssh2/lib/libssh2.1.dylib
libssp.0.dylib -> ../../../../../opt/gcc/lib/gcc/current/libssp.0.dylib
libstdc++.6.dylib -> ../../../../../opt/gcc/lib/gcc/current/libstdc++.6.dylib
```

**Comment by carlocab** [comment]
> On topic of Julia but unrelated to this PR, it may be nice to detect symlinks to other formulae and use that to avoid some "Dependencies with no linkage".

It's nice, but doesn't seem worth the extra code since no other formula does this.

**Comment by cho-m** [comment]
> It's nice, but doesn't seem worth the extra code since no other formula does this.

It would work for things like `ca-certificates` which we always symlink pem file from.

Maybe also `ppsspp` though we skip linkage check as `molten-vk` has an executable.

---

Also related would be header usage (e.g. `boost`, `eigen`, etc) and having a section that details that these deps are in use even though they are not linked libs.

File: Library/Homebrew/linkage_checker.rb [hunk 1]
```diff
@@ -92,6 +92,13 @@ def dylib_to_dep(dylib)
     Regexp.last_match(2)
   end
 
+  sig { params(file: String).returns(T::Boolean) }
+  def broken_dylibs_allowed?(file)
+    return false if formula.name != "julia"
```
Comment (cho-m) [inline Library/Homebrew/linkage_checker.rb:97]: I noticed there is some logic afterward that checks `if formula`. I guess there is a chance formula will not be resolved?
https://github.com/Homebrew/brew/blob/5b0a84bc7a67da90964ba88377b7192e8482c893/Library/Homebrew/linkage_checker.rb#L317-L320

If so, can modify to check `if formula.blank? && ...`

Comment (carlocab) [reply Library/Homebrew/linkage_checker.rb:97]: Could just do
```ruby
if formula&.name != "julia"
```

**Comment by MikeMcQuaid** [comment]
Thanks @cho-m!

**Pull request merged**